packages/components/src/custom-select-control-v2/index.tsx

```diff
@@ -0,0 +1,72 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import * as Ariakit from '@ariakit/react';
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext, useMemo } from '@wordpress/element';
+/**
+ * Internal dependencies
+ */
+import { useCx } from '../utils/hooks/use-cx';
+import * as Styled from './styles';
+import type {
+	CustomSelectProps,
+	CustomSelectItemProps,
+	CustomSelectContext as CustomSelectContextType,
+} from './types';
+
+export const CustomSelectContext =
+	createContext< CustomSelectContextType >( undefined );
+
+export function CustomSelect( props: CustomSelectProps ) {
+	const {
+		children,
+		defaultValue,
+		label,
+		onChange,
+		size = 'default',
+		value,
+		styledValue,
+	} = props;
+
+	const store = Ariakit.useSelectStore( {
+		setValue: ( nextValue ) => onChange?.( nextValue ),
+		defaultValue,
+		value,
+	} );
+
+	const { value: currentValue } = store.useState();
+
+	const cx = useCx();
+
+	const classes = useMemo(
+		() => cx( Styled.inputSize[ size ] ),
+		[ cx, size ]
+	);
+
+	return (
+		<CustomSelectContext.Provider value={ { store } }>
```

Comment:
Nit: it's a good habit to render context providers as "low" in the tree as possible, to avoid unnecessary re-renders. In this case, I would suggest moving it inside `Styled.CustomSelectPopover`, wrapping around `children`

Reply:
Ah, good to know! Thank you for sharing!